File: telemetry/src/service-name.js
"use strict";
/**
 * service.name resolution, shared by tracing and the http logger so spans
 * and http.access lines always agree:
 *
 *   OTEL_SERVICE_NAME > FLY_APP_NAME > app's own package.json name > "unknown-service"
 *
 * The package.json read is best-effort from the process working directory —
 * correct under the fleet convention that CMD execs node from the app root.
 */

const fs = require("node:fs");
const path = require("node:path");

function packageName() {
  try {
    const raw = fs.readFileSync(path.join(process.cwd(), "package.json"), "utf8");
    const name = JSON.parse(raw).name;
    return typeof name === "string" && name.trim() ? name.trim() : undefined;
  } catch {
    return undefined;
  }
}

function resolveServiceName() {
  return process.env.OTEL_SERVICE_NAME || process.env.FLY_APP_NAME || packageName() || "unknown-service";
}

module.exports = { resolveServiceName };

File: telemetry-dotnet/src/Beehive.Telemetry/ServiceName.cs
using System.Reflection;

namespace Beehive.Telemetry;

/// <summary>
/// service.name resolution, shared by tracing, the http logger and the app logger so spans
/// and log lines always agree:
/// <c>OTEL_SERVICE_NAME &gt; FLY_APP_NAME &gt; entry assembly name &gt; "unknown-service"</c>.
/// </summary>
internal static class ServiceName
{
    private static string? cached;

    internal static string Resolve()
    {
        var value = cached;
        if (value is not null)
        {
            return value;
        }

        value = TelemetryEnv.Raw("OTEL_SERVICE_NAME")
            ?? TelemetryEnv.Raw("FLY_APP_NAME")
            ?? EntryAssemblyName()
            ?? "unknown-service";

        cached = value;
        return value;
    }

    private static string? EntryAssemblyName()
    {
        try
        {
            var name = Assembly.GetEntryAssembly()?.GetName().Name;
            return string.IsNullOrWhiteSpace(name) ? null : name.Trim();
        }
        catch (Exception)
        {
            return null;
        }
    }

    /// <summary>Test hook: drops the memoised value so a changed environment is re-read.</summary>
    internal static void ResetForTests() => cached = null;
}
